proof/refine/RISCV64/ArchAcc_R.thy

```diff
@@ -12,19 +12,126 @@ theory ArchAcc_R
+  storePTE.valid_pspace'[simplified valid_obj'_def pred_conj_def, simplified]
+
+end
+
+context begin interpretation Arch . (*FIXME: arch_split*)
 
-lemma corres_gets_asid:
+method readObject_arch_obj_at'_method
+  =  clarsimp simp: readObject_def obind_def omonad_defs split_def loadObject_default_def obj_at'_def
+                    objBits_simps bit_simps' typ_at_to_obj_at_arches
+             split: option.splits if_split_asm
+
+lemma readObject_misc_arch_ko_at'[simp, dest!]:
+  shows
+  readObject_ko_at'_pte: "readObject p s = Some (pte :: pte) \<Longrightarrow> ko_at' pte p s" and
+  readObject_ko_at'_asidpool: "readObject p s = Some (asidp :: asidpool) \<Longrightarrow> ko_at' asidp p s"
+   by readObject_arch_obj_at'_method+
+
+lemma readObject_misc_arch_obj_at'[simp]:
+  shows
+  readObject_pte_at'[simplified]: "bound (readObject p s ::pte option) \<Longrightarrow> pte_at' p s" and
+  readObject_asid_pool_at'[simplified]: "bound (readObject p s ::asidpool option) \<Longrightarrow> asid_pool_at' p s"
+   by readObject_arch_obj_at'_method+
+
+method no_ofail_readObject_method =
+  clarsimp simp: obj_at'_def readObject_def obind_def omonad_defs split_def no_ofail_def,
+  rule ps_clear_lookupAround2, assumption+, simp,
+  blast intro: is_aligned_no_overflow,
+  clarsimp simp: bit_simps' project_inject obj_at_simps lookupAround2_known1 split: option.splits
+
+lemma no_ofail_arch_obj_at'_readObject_pte[simp]:
+  "no_ofail (obj_at' (P::pte \<Rightarrow> bool) p) (readObject p::pte kernel_r)"
+  by no_ofail_readObject_method
+
+lemma no_ofail_arch_obj_at'_readObject_asidpool[simp]:
+  "no_ofail (obj_at' (P::asidpool \<Rightarrow> bool) p) (readObject p::asidpool kernel_r)"
+  by no_ofail_readObject_method
+
+lemma no_ofail_arch_misc_readObject:
+  shows
+  no_ofail_pte_at'_readObject[simp]: "no_ofail (pte_at' p) (readObject p::pte kernel_r)" and
+  no_ofail_asidpool_at'_readObject[simp]: "no_ofail (asid_pool_at' p) (readObject p::asidpool kernel_r)"
+   by (clarsimp simp: typ_at_to_obj_at_arches no_ofail_def
+               dest!: no_ofailD[OF no_ofail_arch_obj_at'_readObject_pte]
+                      no_ofailD[OF no_ofail_arch_obj_at'_readObject_asidpool])+
+
+(* aliases for compatibility with master *)
+
+lemmas getPTE_wp = setObject_pte.get_wp
+lemmas getASID_wp = setObject_asidpool.getObject_wp
+
+lemmas getObject_pte_inv[wp] = setObject_pte.getObject_inv
+lemmas getObject_asidpool_inv = setObject_asidpool.getObject_inv
+
+lemmas get_pte_sp' = setObject_pte.getObject_sp'
+lemmas get_asidpool_sp' = setObject_asidpool.getObject_sp'
+
+lemmas setObject_ASID_cteCaps_of[wp] = ctes_of_cteCaps_of_lift [OF setObject_asidpool.ctes_of]
+lemmas storePTE_cteCaps_of[wp] = ctes_of_cteCaps_of_lift [OF storePTE.ctes_of]
+
+lemma no_0_obj'_abstract:
+  "(s, s') \<in> state_relation \<Longrightarrow> no_0_obj' s' \<Longrightarrow> kheap s 0 = None"
+  by (auto intro: pspace_relation_None simp add: no_0_obj'_def)
+
+declare if_cong[cong]
+
```

Comment:
This is a dangerous declaration, a) for non-termination and b) for general simp performance. We should not mess with if congruence rules globally and I thought we had managed to eliminate that. Did we re-introduce this in `rt`?

In any case, now that you have the theory working, I'd remove this and add only a local `notes if_cong[cong]` to the lemmas that need it (or even just the simp invocations that need it).

Reply:
There is a `declare if_cong[cong]` in RISCV on master. And there's a seemingly different `declare if_cong`in ARM on rt. But sure, I will try to remove it

Reply:
Then I was less successful eradicating it than I thought (maybe I just did AInvs). If it can be removed, that would be really good. If it makes everything unmanageable, it's Ok to put a FIXME on it and leave for now.

Reply:
Yeah a few proofs did break in the CSpace files, too, so I will add the FIXME RT. Is there anything you meant by "tagged"?

Reply:
Cool, `FIXME RT` is good (`RT` is the tag)